File: 45/main.py
import json
from pathlib import Path


class Solution:
    def jump(self, nums: list[int]) -> int:
        max_dist, curr_jumps, curr_end = 0, 0, 0
        for idx, jump in enumerate(nums[:-1]):
            max_dist = max(max_dist, idx+jump)
            if idx == curr_end:
                # I need to make a jump here
                curr_jumps += 1
                curr_end = max_dist

        return curr_jumps


def main():
    files = Path(".").glob("*.json")
    for fname in files:
        input_ = json.load(open(fname))
        nums = input_["nums"]
        print(nums)
        # nums = nums*10000
        out = Solution().jump(nums)
        res = input_["out"]
        print(out)
        assert out == res, (out, res)


if __name__ == "__main__":
    main()

File: 45/main.cpp
#include<vector>
#include<iostream>
using namespace std;
int MAX_INT = 1<<30;

class Solution {
public:
    int jump(const vector<int>& nums) {
        auto dp = vector<int>(nums.size(), MAX_INT);
        dp[0] = 0;
        for(size_t i = 0; i < dp.size(); i++) {
            for(int j = 0; j <= nums[i]; j++) {
                if(i+j >= dp.size()) {
                    break;
                }
                dp[i+j] = min(dp[i+j], dp[i]+1);
            }
        }
        return dp[dp.size()-1];
    }
};

int main() {
    auto sol = Solution();
    cout << sol.jump({2, 3, 1, 1, 4});
    return 0;
}
